File: LeetCode/973.K_Closest_Points_to_Origin/973.K_Closest_Points_to_Origin.java
class Solution {

    class Point {
        int x, y;
        double dis;

        public Point(int x, int y) {
            this.x = x;
            this.y = y;
            this.dis = Math.pow(x, 2) + Math.pow(y, 2);
        }
    }

    public int[][] kClosest(int[][] points, int k) {
        PriorityQueue<Point> maxHeap = new PriorityQueue<>((p1, p2) -> Double.compare(p1.dis, p2.dis));
        for (int[] p : points) {
            maxHeap.offer(new Point(p[0], p[1]));
        }

        int[][] res = new int[k][2];
        for (int i = 0; i < k; i++) {
            Point p = maxHeap.poll();
            res[i][0] = p.x;
            res[i][1] = p.y;
        }
        return res;
    }
}

File: LeetCode/973.K_Closest_Points_to_Origin/973.K_Closest_Points_to_Origin.swift
class Solution {
    struct Point: Comparable {
        let dis: Int
        let idx: Int

        init(_ dis: Int, _ idx: Int) {
            self.dis = dis
            self.idx = idx
        }

        static func < (lhs: Point, rhs: Point) -> Bool { lhs.dis < rhs.dis }
        static func == (lhs: Point, rhs: Point) -> Bool { lhs.dis == rhs.dis }
    }

    func kClosest(_ points: [[Int]], _ k: Int) -> [[Int]] {
        var heap = Heap<Point>()

        for (i, point) in points.enumerated() {
            let dis = point[0] * point[0] + point[1] * point[1]
            heap.insert(Point(dis, i))
            if heap.count > k {
                heap.popMax()
            }
        }

        var res: [[Int]] = []
        for p in heap.unordered {
            res.append(points[p.idx])
        }

        return res
    }
}
